File: report/src/main/java/com/hartwig/actin/report/pdf/tables/IneligibleActinTrialsGenerator.java
package com.hartwig.actin.report.pdf.tables;

import java.util.List;
import java.util.Set;
import java.util.StringJoiner;

import com.google.common.collect.Lists;
import com.hartwig.actin.algo.datamodel.TreatmentMatch;
import com.hartwig.actin.algo.interpretation.EvaluatedTrial;
import com.hartwig.actin.algo.interpretation.EvaluatedTrialExtractor;
import com.hartwig.actin.report.pdf.util.Cells;
import com.hartwig.actin.report.pdf.util.Formats;
import com.hartwig.actin.report.pdf.util.Tables;
import com.itextpdf.layout.element.Table;

import org.apache.logging.log4j.util.Strings;
import org.jetbrains.annotations.NotNull;

public class IneligibleActinTrialsGenerator implements TableGenerator {

    @NotNull
    private final List<EvaluatedTrial> trials;
    @NotNull
    private final String source;
    private final float trialColWidth;
    private final float acronymColWidth;
    private final float cohortColWidth;
    private final float molecularEventColWidth;
    private final float ineligibilityReasonColWith;
    private final float cohortOpenColWidth;

    @NotNull
    public static IneligibleActinTrialsGenerator fromTreatmentMatch(@NotNull TreatmentMatch treatmentMatch, @NotNull String source,
            float contentWidth) {
        List<EvaluatedTrial> ineligibleTrials = Lists.newArrayList();
        for (EvaluatedTrial trial : EvaluatedTrialExtractor.extract(treatmentMatch)) {
            if (!trial.isPotentiallyEligible()) {
                ineligibleTrials.add(trial);
            }
        }

        float trialColWidth = contentWidth / 10;
        float acronymColWidth = contentWidth / 10;
        float cohortColWidth = contentWidth / 4;
        float molecularColWidth = contentWidth / 10;
        float cohortOpenColWidth = contentWidth / 10;
        float ineligibilityReasonColWidth =
                contentWidth - (trialColWidth + acronymColWidth + cohortColWidth + molecularColWidth + cohortOpenColWidth);

        return new IneligibleActinTrialsGenerator(ineligibleTrials,
                source,
                trialColWidth,
                acronymColWidth,
                cohortColWidth,
                molecularColWidth,
                ineligibilityReasonColWidth,
                cohortOpenColWidth);
    }

    private IneligibleActinTrialsGenerator(@NotNull final List<EvaluatedTrial> trials, @NotNull final String source,
            final float trialColWidth, final float acronymColWidth, final float cohortColWidth, final float molecularEventColWidth,
            final float ineligibilityReasonColWith, final float cohortOpenColWidth) {
        this.trials = trials;
        this.source = source;
        this.trialColWidth = trialColWidth;
        this.acronymColWidth = acronymColWidth;
        this.cohortColWidth = cohortColWidth;
        this.molecularEventColWidth = molecularEventColWidth;
        this.ineligibilityReasonColWith = ineligibilityReasonColWith;
        this.cohortOpenColWidth = cohortOpenColWidth;
    }

    @NotNull
    @Override
    public String title() {
        return source + " trials not considered eligible (" + trials.size() + ")";
    }

    @NotNull
    @Override
    public Table contents() {
        Table table = Tables.createFixedWidthCols(trialColWidth,
                acronymColWidth,
                cohortColWidth,
                molecularEventColWidth,
                ineligibilityReasonColWith,
                cohortOpenColWidth);

        table.addHeaderCell(Cells.createHeader("Trial"));
        table.addHeaderCell(Cells.createHeader("Acronym"));
        table.addHeaderCell(Cells.createHeader("Cohort"));
        table.addHeaderCell(Cells.createHeader("Molecular event"));
        table.addHeaderCell(Cells.createHeader("Ineligibility reasons"));
        table.addHeaderCell(Cells.createHeader("Cohort open"));

        for (EvaluatedTrial trial : trials) {
            table.addCell(Cells.createContent(trial.trialId()));
            table.addCell(Cells.createContent(trial.acronym()));
            table.addCell(Cells.createContent(trial.cohort() != null ? trial.cohort() : Strings.EMPTY));
            table.addCell(Cells.createContent("TODO"));
            table.addCell(Cells.createContent(concat(trial.fails())));
            table.addCell(Cells.createContentYesNo(trial.isOpen() ? "Yes" : "No"));
        }

        return Tables.makeWrapping(table);
    }

    @NotNull
    private static String concat(@NotNull Set<String> strings) {
        StringJoiner joiner = Formats.commaJoiner();
        for (String string : strings) {
            joiner.add(string);
        }
        return Formats.valueOrDefault(joiner.toString(), "?");
    }
}

File: report/src/main/kotlin/com/hartwig/actin/report/pdf/tables/trial/IneligibleActinTrialsGenerator.kt
package com.hartwig.actin.report.pdf.tables.trial

import com.hartwig.actin.datamodel.trial.TrialSource
import com.hartwig.actin.report.interpretation.InterpretedCohort
import com.hartwig.actin.report.pdf.tables.trial.TrialGeneratorFunctions.addTrialsToTable
import com.hartwig.actin.report.pdf.util.Cells
import com.hartwig.actin.report.pdf.util.Tables
import com.hartwig.actin.report.pdf.util.Tables.makeWrapping
import com.itextpdf.layout.element.Table

private const val SMALL_PADDING_DISTANCE = 0.1f
private const val NORMAL_PADDING_DISTANCE = 1f

class IneligibleActinTrialsGenerator(
    private val cohorts: List<InterpretedCohort>,
    private val requestingSource: TrialSource?,
    private val title: String,
    private val footNote: String?,
    private val trialColWidth: Float,
    private val subTableWidths: FloatArray,
    private val includeIneligibilityReasonCol: Boolean,
    private val paddingDistance: Float
) : ActinTrialsGenerator {

    override fun title(): String {
        return title
    }

    override fun contents(): Table {
        val table = Tables.createFixedWidthCols(trialColWidth, subTableWidths.sum())
        if (cohorts.isNotEmpty()) {
            table.addHeaderCell(Cells.createContentNoBorder(Cells.createHeader("Trial")))
            val headerSubTable = Tables.createFixedWidthCols(*subTableWidths)
            sequenceOf("Cohort", "Molecular", "Sites", "Ineligibility reasons").map(Cells::createHeader)
                .forEach(headerSubTable::addHeaderCell)
            table.addHeaderCell(Cells.createContentNoBorder(headerSubTable))
        }
        addTrialsToTable(
            cohorts,
            externalTrials = emptySet(),
            requestingSource,
            homeCountry = null,
            table,
            subTableWidths,
            InterpretedCohort::fails,
            includeIneligibilityReasonCol,
            paddingDistance
        )

        if (footNote != null) {
            table.addCell(Cells.createSpanningSubNote(footNote, table))
        }
        return makeWrapping(table)
    }

    override fun getCohortSize(): Int {
        return cohorts.size
    }

    companion object {
        fun forCohorts(
            cohorts: List<InterpretedCohort>, requestingSource: TrialSource?, width: Float, openOnly: Boolean = false
        ): IneligibleActinTrialsGenerator {
            val ineligibleCohorts = cohorts.filter { !it.isPotentiallyEligible && (it.isOpen || !openOnly) }
            val (trialColWidth, subTableWidths) = getColumnWidths(width, true)
            val title =
                "Trials and cohorts that are considered ineligible (${ineligibleCohorts.size})"
            val footNote = if (!openOnly) "Closed cohorts are shown in grey." else null
            return IneligibleActinTrialsGenerator(
                ineligibleCohorts,
                requestingSource,
                title,
                footNote,
                trialColWidth,
                subTableWidths,
                true,
                paddingDistance = NORMAL_PADDING_DISTANCE
            )
        }

        fun forNonEvaluableAndIgnoredCohorts(
            ignoredCohorts: List<InterpretedCohort>,
            nonEvaluableCohorts: List<InterpretedCohort>,
            requestingSource: TrialSource?,
            width: Float,
        ): IneligibleActinTrialsGenerator {
            val nonEvaluableAndIgnoredCohorts = ignoredCohorts + nonEvaluableCohorts
            val (trialColWidth, subTableWidths) = getColumnWidths(width, false)
            val title =
                "Trials and cohorts that are not evaluable or ignored (${nonEvaluableAndIgnoredCohorts.size})"
            return IneligibleActinTrialsGenerator(
                nonEvaluableAndIgnoredCohorts,
                requestingSource,
                title,
                footNote = null,
                trialColWidth,
                subTableWidths,
                false,
                paddingDistance = SMALL_PADDING_DISTANCE
            )
        }

        private fun getColumnWidths(
            width: Float, includeIneligibilityReason: Boolean = false
        ): Pair<Float, FloatArray> = width.let { w ->
            val trialWidth = if (includeIneligibilityReason) w / 9 else w / 4
            val cohortWidth = if (includeIneligibilityReason) w / 4 else w / 2
            val molecularWidth = if (includeIneligibilityReason) w / 7 else w / 4
            val sitesWidth = if (includeIneligibilityReason) w / 7 else w / 4
            val remainingWidth = w - (trialWidth + cohortWidth + molecularWidth + sitesWidth)
            trialWidth to listOfNotNull(
                cohortWidth,
                molecularWidth,
                sitesWidth,
                remainingWidth.takeIf { includeIneligibilityReason }
            ).toFloatArray()
        }
    }
}